### PR TITLE
Fix recaptcha detection on iOS.

### PR DESCRIPTION
**Pull request by rammie**
By default, iOS automatically redirects all the way to the google recaptcha page. This change prevents that from happening and instead uses the servers incapsula detection to trigger displaying the captcha instead (without request blocking). This makes it the implementation behave the same way as Android, and captcha solving works as expected.

### CODE REVIEW

**Comment by lippytak** [comment]
@rammie I don't think I can test this since PR builds don't go to TestFlight. LMK if you can manually trigger a TestFlight build or if you merge it into master I can QA.

File: src/ios/CDVLocalTunnel.m [hunk 1]
```diff
@@ -747,11 +747,23 @@ - (BOOL)webView:(UIWebView*)theWebView shouldStartLoadWithRequest:(NSURLRequest*
         return NO;
     }
     else if (requestUrl != nil && ![url isEqual:requestUrl] && ![[url absoluteString] isEqualToString:@"about:blank"]) {
```
Comment (paetling) [inline src/ios/CDVLocalTunnel.m:750]: can we add a comment above this `else if` that this is for when the url redirects in the webview? That was not clear to me from reading so I got a bit confused when trying to poke around in this issue.

File: src/ios/CDVLocalTunnel.m [hunk 2]
```diff
@@ -747,11 +747,23 @@ - (BOOL)webView:(UIWebView*)theWebView shouldStartLoadWithRequest:(NSURLRequest*
...
+            NSLog(@"---- DETECTED REDIRECT TO RECAPTCHA.");
+            enableRequestBlocking = false;
+            NSArray* cookies = [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookiesForURL:requestUrl];
+            NSDictionary* cookieHeader = [NSHTTPCookie requestHeaderFieldsWithCookies:cookies];
+            NSString* cookieStr = [cookieHeader objectForKey:@"Cookie"];
+            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
+                                                      messageAsDictionary:@{@"type":@"requestdone", @"url":[url absoluteString], @"cookies":cookieStr}];
+            [pluginResult setKeepCallback:[NSNumber numberWithBool:YES]];
+            [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
+            return NO;
```
Comment (paetling) [inline src/ios/CDVLocalTunnel.m:764]: Hmm trying to parse the logic here and got a bit confused. 

Things I think I know:
* `shouldStartLoadWithRequest` controls whether or not the webview will try and load a URL
* For the android localtunnel, we try and load the recaptcha page first, get a "You are a robot" error and then try and launch the recaptcha page in its own new captcha webview
* This code is sending back a request done with the google url and choosing not to load the recpatcha page. 

Is the reason why we are returning here instead of loading is because the page should contain the "robots" html that the server is looking for and thus recaptcha should happen as expected?

Is the reason it was not working before is because the recaptcha page was loading but was loading in the non visible webview? Even if you forced the webview to be visible, the localtunnel.ts file does not have a callback setup for recaptcha being done, so it still would not work (behavior jake is seeing)?

Comment (rammie) [reply src/ios/CDVLocalTunnel.m:764]: > Is the reason why we are returning here instead of loading is because the page should contain the "robots" html that the server is looking for and thus recaptcha should happen as expected?

Correct.

> Is the reason it was not working before is because the recaptcha page was loading but was loading in the non visible webview? Even if you forced the webview to be visible, the localtunnel.ts file does not have a callback setup for recaptcha being done, so it still would not work (behavior jake is seeing)?

Almost correct. The reason we failed to solve it with show webview enabled in the debugger was because request blocking was enabled at the time. So several XHR requests made by the recaptcha page were failing.

**Review by paetling — APPROVED** [review]
Ahh. These fixes make sense. Nice work!

I would like to advocate for a bit more why comments especially in the inappbrowser plugins. It is really hard to parse why each code block exists right now when reading through the first few times. 

For both this code and the code you just added for OK fix stuff I would like to advocate for some comments explaining what you are doing. 

For this code something like: `# We do not allow the recaptcha page to load because we need to insular (you are a robots) page to be returned to the server for our recaptcha flow to fully work. Simply redirecting to recpatcha breaks this pattern`. 

For the OK code something like: `# Reset password page is correctly loading but an erroneous error is being thrown by iOS. That leads to us not loading the page properly in the webview. Instead we just capture the error on OK password reset and say everything succeeded`
